### DISCUSSION TITLE
Ongoing work to support Kamstrup 382M meters (mainly relevant for Denmark)

### DISCUSSION THREAD

**Discussion by ArnieO** [UtilitechAS/amsreader-firmware#725] [General]
### For some time now we have been working on finding a way to read Kamstrup 382M meters. This thread will provide information on the status and progress on that work.

Our [Pow-K](https://www.amsleser.no/hjem/1-11-pow-k-kamstrup.html) can read data from all Kamstrup Omnipower meters, that can be configured to send data to the HAN port every 10 seconds:
![image](https://github.com/UtilitechAS/amsreader-firmware/assets/10295178/79e5676b-012b-4999-a16f-85a2dd2b1804)

A number of users in Denmark have an older type Kamstrup meters that cannot be configured to push data to the HAN-port, this seems to be mainly the Kamstrup 382M models:
![image](https://github.com/UtilitechAS/amsreader-firmware/assets/10295178/9a7f94c2-54c2-4e9e-a3a9-30a525852845)

We do not have full overview which grid companies in Denmark use the 382M models, we know of these:
- Dinel (https://dinel.dk) 
- Veksel (https://www.veksel.dk)

Please respond in this thread if you have a Kamstrup 382M with a different grid company.

In order to read data from the 382M model, the connected equipment needs to establish communication with the meter and actively request it to send data, i.e. do a "data pull". The protocol used for establishing such communication is a Kamstrup proprietary protocol, called "KMP protocol".

As we are dedicated to keeping our firmware code available to our users, we have over the last months had a discussion with Kamstrup on how to implement the "Data pull" and still publish the source code. The conclusion was that in order to obtain access to the documentation for the KMP protocol, we would need to sign a Non-Disclosure Agreement (NDA), where we commit to not publishing source code.

So we looked for a workaround. Using the (non-proprietary) DLMS protocol seemed to be a promising alternative. However, after testing that for a while we saw that in order to access the actual measurement data, a special DMLS encryption key is needed, which is not the encryption key the grid companies can provide to their customers. So using DLMS turned out to be a dead end: It was Back to square one!

We then figured that the "second best" solution will be to use the KMP protocol, but publish our code in such a way that the code pieces where the KMP protocol is implemented is only published as binary - and then assembled into the binary file in some way during the code build. Please note that I am not the one in our team that knows the details about this, so the explanation may be slightly inaccurate. But the point is that we can find a way of implementing the KMP protocol while publishing all other source code. 

So we recently signed the NDA with Kamstrup - and have now received the KMP protocol documentation. Next step will be writing the necessary code - and testing it. 

We do not have access to a Kamstrup 382 meter, so we will rely on helpers in Denmark to whom we will send prototype boards that are modified to be able to communicate with those meters. For the time being we believe we have the candidates we need - but will inform here if we need additional helpers. (We used the same method when we developed Pow-P1 without having laid hand on a single P1 meter yet - and that worked quite well.)

Our current objective is to have solved this sometime in Q2/24. We will then also have next generation boards available for sale, where the necessary modification is done (it's just a question of routing an available microcontroller GPIO to a pin on the connector that connects the board to the meter).

**Comment by erathor82** [comment]
Excellent news! I have one of those meters in denmark. Our grid company name is Veksel. Let me know if you need further testers or when it's released for purchase. 👍

**Reply by ArnieO** [reply]
Thank you for letting me know regarding Veksel, I'll update the posting with that. 👍 

I'll notify here if we need more testers! 
Please follow the thread to get updates regarding progress including release for purchase.

**Comment by Sonnengruesser** [comment]
That sounds very promising! 
Is there still a key required when using KMP? If yes what's the name so I can guide my energy provider to request the correct one at Kamstrup.

I volunteer as a tester in Austria - if needed.

**Reply by ArnieO** [reply]
So the Kamstrup 382M is also used in Austria? That is interesting news for us, it means that our market is potentially larger than we knew. 

The encryption key used on these meters should be the one called GPK11 (single key).

Thank you for volunteering to test, but for the moment we believe we have what we need!

**Reply by Sonnengruesser** [reply]
No, sorry, wrong information.
I have an early model of the Omnipower which doesn't support data push. Hope the 382M KMP solution also works.

**Reply by ArnieO** [reply]
If we have understood the situation correctly, all Omnipower can be configured for Data push. The problem is sometimes that the grid company do not have access to the remote config service (maybe a cost question). We have two customers with small grid companies in Switzerland (I do not know which model Omnipower) where the grid company sent a technician to reconfigure via the optical port, using a laptop with the necessary device on a cable.

**Reply by bommel947** [reply]
> No, sorry, wrong information. I have an early model of the Omnipower which doesn't support data push. Hope the 382M KMP solution also works.

Hi, it looks like we have the same issue. I am also struggling to receive data from my omnipower smartmeter. 
I have tried the KMP solution without any luck. How to identify the model of the omnipower startmeter??

**Reply by ArnieO** [reply]
> How to identify the model of the omnipower startmeter??

If you post a photo of the meter, I'll help you identify it. 
You may want to mask out the serial number.

**Reply by bommel947** [reply]
This it my meter.
Thanks for helping me out !
![omnipower](https://github.com/user-attachments/assets/2f82b3f9-4829-4985-aaac-934fc13f80b9)

**Reply by ArnieO** [reply]
This is a normal Kamstrup Omnipower meter, and you are in Austria with https://www.kwg.at/

I have just finished a discussion with an other KWG-customer in Austria that struggled to get his meter to deliver data. 
He sent KWG an email yesterday, and requested them to do this:
1) Activate the HAN port
2) Configure the port for "HAN push" (the meter will then send data each 10 second). 

And he now has everything up and running. 

**Number 2) is important**, as it seems like [KWG has previously proposed Smart-Me units](https://www.kwg.at/smart-me/) to use on these meters, and we think they then configured the meters for "Data pull". I believe Smart-Me no longer sells outside of Switzerland, so it is a bit strange that KWG still advertise that solution. 

Anyway - please contact KWG with the indicated request.

**Comment by straend** [comment]
This meter is also in use in Finland (at least here in the southern part). Can also try to do some testing if needed.

**Reply by ArnieO** [reply]
Hi @straend , that is interesting information! Thank you for volunteering to test, we will start working with two persons in Denmark - and maybe extend the number of testers later on.
Current status is that our two helpers in Denmark has received their hardware - and are waiting for us to issue the first test firmware version. So - moving forward slowly but surely.

**Reply by straend** [reply]
Good to hear, hoping it goes well.

This is the meter we have.
![Screenshot from 2024-02-19 18-58-04](https://github.com/UtilitechAS/amsreader-firmware/assets/1004231/189130fe-0d90-42a1-99e9-3b7366cda6fd)

**Comment by kennethslot** [comment]
> ### For some time now we have been working on finding a way to read Kamstrup 382M meters. This thread will provide information on the status and progress on that work.
> Our [Pow-K](https://www.amsleser.no/hjem/1-11-pow-k-kamstrup.html) can read data from all Kamstrup Omnipower meters, that can be configured to send data to the HAN port every 10 seconds: ![image](https://private-user-images.githubusercontent.com/10295178/300267085-79e5676b-012b-4999-a16f-85a2dd2b1804.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgzNTkwOTAsIm5iZiI6MTcwODM1ODc5MCwicGF0aCI6Ii8xMDI5NTE3OC8zMDAyNjcwODUtNzllNTY3NmItMDEyYi00OTk5LWExNmYtODVhMmRkMmIxODA0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE5VDE2MDYzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNhZjUzODJhODJmNDVjOGIzNmFiY2NmY2VhY2ZjOWEzYTcyNmFjYTc3MTUyYWJkMjg5Yzg0MmE3N2NkZGMxOWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.xQMN7I3NqMEtfQ2Evl9FuVCna18K4ULSXAVtt5uATa8)
> 
> A number of users in Denmark have an older type Kamstrup meters that cannot be configured to push data to the HAN-port, this seems to be mainly the Kamstrup 382M models: ![image](https://private-user-images.githubusercontent.com/10295178/300266852-9a7f94c2-54c2-4e9e-a3a9-30a525852845.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgzNTkwOTAsIm5iZiI6MTcwODM1ODc5MCwicGF0aCI6Ii8xMDI5NTE3OC8zMDAyNjY4NTItOWE3Zjk0YzItNTRjMi00ZTllLWEzYTktMzBhNTI1ODUyODQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE5VDE2MDYzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEzZTQxYTQxMTVlZDI4ZjFjMWJhY2FmNzEwNmI3NDkzZmNkODQyYmEyN2UzMzkxOGJjOWQ5MTBlMGY3ZTI0YTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3Fq445l9tobSsWe31ky7DKke1xyKHd5Gqzg7sC2Vj14)
> 
> We do not have full overview which grid companies in Denmark use the 382M models, we know of these:
> 
> * Dinel (https://dinel.dk)
> * Veksel (https://www.veksel.dk)
> 
> Please respond in this thread if you have a Kamstrup 382M with a different grid company.
> 
> In order to read data from the 382M model, the connected equipment needs to establish communication with the meter and actively request it to send data, i.e. do a "data pull". The protocol used for establishing such communication is a Kamstrup proprietary protocol, called "KMP protocol".
> 
> As we are dedicated to keeping our firmware code available to our users, we have over the last months had a discussion with Kamstrup on how to implement the "Data pull" and still publish the source code. The conclusion was that in order to obtain access to the documentation for the KMP protocol, we would need to sign a Non-Disclosure Agreement (NDA), where we commit to not publishing source code.
> 
> So we looked for a workaround. Using the (non-proprietary) DLMS protocol seemed to be a promising alternative. However, after testing that for a while we saw that in order to access the actual measurement data, a special DMLS encryption key is needed, which is not the encryption key the grid companies can provide to their customers. So using DLMS turned out to be a dead end: It was Back to square one!
> 
> We then figured that the "second best" solution will be to use the KMP protocol, but publish our code in such a way that the code pieces where the KMP protocol is implemented is only published as binary - and then assembled into the binary file in some way during the code build. Please note that I am not the one in our team that knows the details about this, so the explanation may be slightly inaccurate. But the point is that we can find a way of implementing the KMP protocol while publishing all other source code.
> 
> So we recently signed the NDA with Kamstrup - and have now received the KMP protocol documentation. Next step will be writing the necessary code - and testing it.
> 
> We do not have access to a Kamstrup 382 meter, so we will rely on helpers in Denmark to whom we will send prototype boards that are modified to be able to communicate with those meters. For the time being we believe we have the candidates we need - but will inform here if we need additional helpers. (We used the same method when we developed Pow-P1 without having laid hand on a single P1 meter yet - and that worked quite well.)
> 
> Our current objective is to have solved this sometime in Q2/24. We will then also have next generation boards available for sale, where the necessary modification is done (it's just a question of routing an available microcontroller GPIO to a pin on the connector that connects the board to the meter).

Hi im using the same 382mxg kampstrup and is also Home assistant user. I am on the N1 grid
please let me know if you need help to test the kmp model

**Reply by ArnieO** [reply]
Hi @kennethslot : So since you ask: N1.dk also has 382M meters. 
That is an interesting reminder, as this is a large grid company covering a large part of Jylland. 

I visited their website this morning to check if they use 382M meters, and from this page I only see Omnipower (when I click the button in teh Kamstrup field). https://n1.dk/elnettet/elmaaler

But that apparently does not tell the full story.

**Comment by ArnieO** [comment]
> Hi im using the same 382mxg kampstrup and is also Home assistant user. I am on the N1 grid
> please let me know if you need help to test the kmp model

Thank you for letting us know, @kennethslot !

**Comment by Sonnengruesser** [comment]
I'm wondering - is there any progress on this topic?

**Reply by ArnieO** [reply]
Slow progress due to recent workload with release and following patching of v2.3.

Two testers in Denmark are equipped with boards - and the initial attempts to communicate with their meters have been done.

**Comment by ArnieO** [comment]
We have made great progress lately!
Two happy testers in Holstebro are now receiving data from their Kamstrup 382M meters:
![image](https://github.com/UtilitechAS/amsreader-firmware/assets/10295178/cc09e33e-454a-4cd8-b778-78d2835e5190)

There are some minor issues left to sort out, so we are not yet ready to release the code - but we're getting very close.

**Reply by ArnieO** [reply]
Ping @chmadkk

**Reply by tauchmi** [reply]
Will this also work with newer Kamstrup meters if Push is not active?

**Reply by ArnieO** [reply]
Thank you for asking - as this is a good question. 

In principle Yes, but we would like to check available information first, and then have a "pilot customer" to try it before we announce that we can support it.

Most grid companies can reconfigure Kamstrup Omnipower meters to Data push if you ask them, but we know that not all small grid companies have that ability.

- What model is this? (A photo of the front would be great!)
- I'm guessing: Austria?
- If so, I expect it to be encrypted, and should then be GPK11 encryption key.
- Link to grid company web site with information regarding the meter would be great. (I can read "technical German", but have some difficulties navigating German websites, so please point to the right area or document!)

**Reply by tauchmi** [reply]
We already talked about this in this issue thread: https://github.com/UtilitechAS/amsreader-firmware/issues/126#issuecomment-1861670898 see above for picture.

Yes, Austria. My local energy provider was not able to answer any questions regarding encryption. I wrote to info-at@kamstrup.com and received answers from service@kamstrup.dk that with the current firmware push is not available and will not be as an upgrade is not possible.

There is no technical information on my grid companies homepage concerning the smart meter.

**Reply by ArnieO** [reply]
💡 Sorry for not having recognized you!

I tried now on my Omnipower the test firmware currently running on the 382M test subjects in Denmark. But it does NOT run on my Omnipower. It should absolutely be possible, but apparently not in the same way as on the 382M.

We will finalize the coding for 382M first, then see if we gan get through also on Omnipower meters.

**Reply by Sonnengruesser** [reply]
I follow this thread too, since I'm at the same position. And I have a very early 684-xxx model too (not a 382M as previously thought).

I requested again the GPK11 key at my small energy provider - my contact there will forward the request to Kamstrup (Kamstrup Austria I guess). 
@tauchmi if you are lucky and have a name at Kamstrup Austria please let me know so I can assist my energy provider.

**Reply by Sonnengruesser** [reply]
I received the GPK11, GPK60 and GPK61 keys although the energy provider had confirmed earlier that my 684-xxx model has a very old firmware that doesn't support data push. However I gave it a try with the KMP software.

When selecting data push with/without keys nothing happens as expected.
When selecting KMP the HAN indication on the webpage turns green, but there is no more output to the serial debug window and no data shows up on the webpage. Ticking "encryption" checkbox doesn't change anything.

@ArnieO is there any way to see which information is sent/received via KMP, for example general (unencrypted) meter data on the info page?

Further is it correct to input the GPK11 into the first line "encryption" and leaving "authentification" empty?

PS: I also got info for a GPK8 key. If this try doesn't work I'm gonna change the energy provider just for the sake of getting a new meter...

**Reply by ArnieO** [reply]
We do not have documentation on the 684-xxx model, so we do not know whether it is possible to read from it with Data pull. But please try!

> When selecting data push with/without keys nothing happens as expected.

Which bitrate settings did you use - did you try this? https://wiki.amsleser.no/en/pow-k#kamstrup-382m-configuration

> @ArnieO is there any way to see which information is sent/received via KMP, for example general (unencrypted) meter data on the info page?

You must use [Telnet debug](https://www.amsleser.no/module/ets_blog/blog?id_post=24) to see the raw data from the meter. 



> Further is it correct to input the GPK11 into the first line "encryption" and leaving "authentification" empty?

Yes, correct.

And we do not know what the GPK8 key is.

**Comment by chmadkk** [comment]
Thats great. Looking forward to purchasing an board with the support.Den 4. jun. 2024 kl. 13.20 skrev ArnieO ***@***.***>:﻿
Ping @chmadkk

**Comment by tjay78** [comment]
Nice to see this finally being integrated into this project :)
I gave up on it a few years ago, but i got it working in ESPHome instead and it's been rock stable for a few years now, but i'll try this when it's ready for testing 👍

**Comment by gskjold** [comment]
If you would like to test this feature, Install version below, go to Configuration -> Meter -> Communication and select KMP. Then set baud and parity, most likely 9600 8N1

[esp32s3.zip](https://github.com/user-attachments/files/15685394/esp32s3.zip)
[esp32solo.zip](https://github.com/user-attachments/files/15685395/esp32solo.zip)
[esp32.zip](https://github.com/user-attachments/files/15685396/esp32.zip)
[esp32c3.zip](https://github.com/user-attachments/files/15685400/esp32c3.zip)
[esp8266.zip](https://github.com/user-attachments/files/15685401/esp8266.zip)
[esp32s2.zip](https://github.com/user-attachments/files/15685402/esp32s2.zip)

**Reply by tjay78** [reply]
Hi 
I don't know what's wrong
I have 3 different ESP32 which run AMS v2.3.5 fine but when I update them on the web interface they boot back to v2.3.5 and if I use esptool they don't boot afterwards??
I don't know if there is something wrong with those files or I am doing something wrong.
I found an old esp8266 that seems to work but unfortunately it is very unstable so I can't test it on it.

**Reply by gskjold** [reply]
I see now that the firmware for vanilla esp32 is larger than the standard partition size, so it makes sense that it wont run. Will look into options at some point.

**Comment by coderegiondk** [comment]
Hi everyone,
I have a Kamstrup 382M meter and is connected to Flow Net (City: Svendborg)
From the communication with them it seems they 're using Kamstrup 382M in the whole net area.
I'll much appreciate all your work and hope for a solution to the problem.

**Reply by ArnieO** [reply]
Thank you for encouraging words!
The code is now running as it should with our testers in Holstebro:
![image](https://github.com/UtilitechAS/amsreader-firmware/assets/10295178/ef2c6dcc-411a-4299-a60c-b80ae6b544fe)

The key to making this work is using the proprietary Kamstrup KMP protocol to communicate with the meter, basically requesting all parameters one by one ("data pull"). As this is a proprietary protocol, we have signed an NDA (Non-disclosure agreement) with Kamstrup regarding its use. Under the clauses of that NDA, we cannot publish the source code for the KMP sections of our code. This creates some extra work for us, as we want to include it as part of the code (published on this Github site) - and not a separate non-published version. The KMP part of the code must be published as binary; not source code. 

This means extra work before we can release the new version of the code that supports Kamstrup 382M. 

**However**
Those of you that are eager to connect to your 382M are welcome to order a [Pow-K](https://lectronz.com/products/pow-k-han-port-reader-for-kamstrup-meters). If you include a note with the order stating that it will be used on a 382M, we will deliver it with "test version" code that can read 382M. It is also fairly easy to upgrade a Pow-K if you get your hands on one - but it must be the latest board layout (v1.6.1). Older Pow-K boards are not compatible with 382M.

**Reply by tjay78** [reply]
HI
it's funny, I'm also from Svendborg and I've had long discussions with both Flow NET and Kamstrup about this
but it seems like, the people you talk to don't really understand what you're asking about and then you don't really get any further..
So I gave up and made it work in ESPhome and Home assistant instead :)

But I'm looking forward to seeing it work on AMSreader..

**Reply by ArnieO** [reply]
I don't think we can expect the average grid company to have that level of technical knowledge regarding the power meters, especially the small grid companies.

We have worked on our hardware and firmware for several years, and have managed to find a technical contact at Kamstrup that knows his stuff - which is why we know both what the issue is and found a way to get around it by using the proprietary KMP protocol. 

The core difficulty is that the 382M cannot be reconfigured to "Data Push" mode. The Omnipower can be configured to "Data Push" either remotely or on-site via the optical port.

"Data Push" mode is when the meter sends a "datagram" (containing a predefined set of parameters) by itself every x seconds (10 seconds for the Kamstrup Omnipower). The alternative is "Data pull", where you establish a bidirectional communication with the meter, and request the various parameters one by one. Which is significantly more challenging in the firmware, as you need to understand and implement the handshake protocol.

Before signing NDA with Kamstrup to be allowed to use KMP, we attempted to get around it by using the open DLMS protocol. But it turned out that there is a password needed to get out anything but trivial information. And that password is not something the grid company has easily available. So we had to abandon that path - and went for Kamstrup KMP.

Then there was of course the additional challenge that we (working out of Norway) do not to have direct access to a 382M meter. So we allied with helpers in Denmark that received prototype boards modified to be like the board version we're currently selling. They have received numerous versions of test firmware, until it is now working as it should.

We now have to find a way to publish source code for the rest of the code, while "hiding" the source code for KMP - and still make it feasible for a hobbyist in Denmark to compile the code for hers/his power meter. It is fully feasible, we just have to find the best way to do it - now that we have working code up and running.

**Reply by carlreid** [reply]
> If you include a note with the order stating that it will be used on a 382M, we will deliver it with "test version" code that can read 382M. 

Just ordered! So glad I came across this. Was looking into it back in Jan 2023, and didn't want the optical eye measuring set up. Today I thought I'd give in, and then find this amazing progress 👍 Looking forward to setting it up.

**Comment by coderegiondk** [comment]
After receiving a PowK+ for my Kamstrup 382m meter I finally are able to get data from the meter.
By the way; My net company is Flow NET in Svendborg.
This project is truly awesome.

**Reply by ArnieO** [reply]
Excellent - and please spread the word around you!

We have contacted the grid companies in Denmark that use 382M, but they all say that due to their monopolist role they cannot post information about our solution on their web site - which means we need some help to spread the information.

**Comment by chmadkk** [comment]
Still no possibility to release a precompiled software with kmp support ? And keeping out on releasing the Source code ?

**Reply by ArnieO** [reply]
That should absolutely be possible - when @gskjold gets an opportunity. Might take a bit of time right now due to vacation. 

Which micro controller?

**Comment by chmadkk** [comment]
Esp32 standard. Or similar.

**Comment by kennethslot** [comment]
Hi Arnie
I think that alle new installations uses the omnipower, but mine is 382m
meter.


Den tirs. 4. jun. 2024 kl. 15.57 skrev ArnieO ***@***.***>:

> Hi @kennethslot <https://github.com/kennethslot> : So since you ask:
> N1.dk also has 382M meters.
> That is an interesting reminder, as this is a large grid company covering
> a large part of Jylland.
>
> I visited their website this morning to check if they use 382M meters, and
> from this page I only see Omnipower (when I click the button in teh
> Kamstrup field). https://n1.dk/elnettet/elmaaler
>
> But that apparently does not tell the full story.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/UtilitechAS/amsreader-firmware/discussions/725#discussioncomment-9665344>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOFLKQNJHYEXWMFT4AMLKQ3ZFXBVPAVCNFSM6AAAAABCOHIRIWVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TMNRVGM2DI>
> .
> You are receiving this because you were mentioned.Message ID:
> <UtilitechAS/amsreader-firmware/repo-discussions/725/comments/9665344@
> github.com>
>


-- 
Venlig Hilsen

Kenneth Slot
Møllevej 1
9240 Nibe
eMail: ***@***.***
Tlf. +45 22 76 39 53
Skypename: kennethslot
Ham: OZ5KSL

**Reply by ArnieO** [reply]
Hi @kennethslot,
OK - that makes sense.

**Comment by gskjold** [comment]
Here is the latest build that includes KMP support:
[esp32s3.zip](https://github.com/user-attachments/files/16180502/esp32s3.zip)
[esp32solo.zip](https://github.com/user-attachments/files/16180503/esp32solo.zip)
[esp32.zip](https://github.com/user-attachments/files/16180504/esp32.zip)
[esp8266.zip](https://github.com/user-attachments/files/16180506/esp8266.zip)
[esp32c3.zip](https://github.com/user-attachments/files/16180507/esp32c3.zip)
[esp32s2.zip](https://github.com/user-attachments/files/16180509/esp32s2.zip)

Note that vanilla ESP32 and ESP32-C3 have been compiled with a custom partition table. If you are upgrading from a previous installation you should back up config and do a full flash including the partitions.bin file.

**Reply by straend** [reply]
Can report that it works here in Finland. 
But the esp32-c3 build seems to lack KMP protocol support. Works great with esp32-s3

**Reply by straend** [reply]
esp8266 with Software Serial was not fast enough for the parsing

**Reply by ArnieO** [reply]
> Can report that it works here in Finland. But the esp32-c3 build seems to lack KMP protocol support. Works great with esp32-s3

So there are Kamstrup 382M meters installed in Finland? That is interesting to know. Which grid company is this?

> esp8266 with Software Serial was not fast enough for the parsing

Good to know, thank you for that information.

**Reply by straend** [reply]
In this case its [Keravan Energia](https://www.keravanenergia.fi/en/home/)

**Reply by ArnieO** [reply]
Yes, I find Kamstrup 382M among their user manuals for power meters:
https://evermade-keravanenergia-website.s3.eu-north-1.amazonaws.com/wp-content/uploads/2021/01/08104158/Kamstrup-lukuohje_a.pdf
